File: src/main/java/com/everyonewaiter/global/exception/GlobalExceptionHandler.java
package com.everyonewaiter.global.exception;

import com.everyonewaiter.domain.shared.AccessDeniedException;
import com.everyonewaiter.domain.shared.AuthenticationException;
import com.everyonewaiter.domain.shared.BusinessException;
import com.everyonewaiter.domain.shared.ErrorCode;
import com.everyonewaiter.global.logging.ExceptionLogger;
import feign.FeignException;
import jakarta.servlet.http.HttpServletRequest;
import java.util.Objects;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.web.HttpMediaTypeNotSupportedException;
import org.springframework.web.HttpRequestMethodNotSupportedException;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.MissingServletRequestParameterException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.context.request.async.AsyncRequestTimeoutException;
import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException;
import org.springframework.web.servlet.resource.NoResourceFoundException;

@RestControllerAdvice
class GlobalExceptionHandler {

  @ExceptionHandler({IllegalArgumentException.class, IllegalStateException.class})
  public ResponseEntity<ErrorResponse> handleIllegal(
      HttpServletRequest request,
      Exception exception
  ) {
    ErrorCode errorCode = ErrorCode.CLIENT_BAD_REQUEST;
    String message = Objects.requireNonNullElse(exception.getMessage(), errorCode.getMessage());
    ExceptionLogger.warn(request, errorCode, message, exception);
    return ResponseEntity.status(errorCode.getStatus())
        .body(ErrorResponse.of(errorCode, message));
  }

  @ExceptionHandler
  public ResponseEntity<ErrorResponse> handleMethodArgsTypeMismatch(
      HttpServletRequest request,
      MethodArgumentTypeMismatchException exception
  ) {
    ErrorCode errorCode = ErrorCode.INVALID_PARAMETER;
    ExceptionLogger.warn(request, errorCode, exception);
    return ResponseEntity.status(errorCode.getStatus())
        .body(ErrorResponse.from(errorCode));
  }

  @ExceptionHandler
  public ResponseEntity<ErrorResponse> handleMissingServletRequestParameter(
      HttpServletRequest request,
      MissingServletRequestParameterException exception
  ) {
    ErrorCode errorCode = ErrorCode.MISSING_PARAMETERS;
    ExceptionLogger.warn(request, errorCode, exception);
    return ResponseEntity.status(errorCode.getStatus())
        .body(ErrorResponse.from(errorCode));
  }

  @ExceptionHandler
  public ResponseEntity<ErrorResponse> handleHttpMessageNotReadable(
      HttpServletRequest request,
      HttpMessageNotReadableException exception
  ) {
    ErrorCode errorCode = ErrorCode.INVALID_REQUEST_BODY;
    ExceptionLogger.warn(request, errorCode, exception);
    return ResponseEntity.status(errorCode.getStatus())
        .body(ErrorResponse.from(errorCode));
  }

  @ExceptionHandler
  public ResponseEntity<ErrorResponse> handleHttpMediaTypeNotSupported(
      HttpServletRequest request,
      HttpMediaTypeNotSupportedException exception
  ) {
    ErrorCode errorCode = ErrorCode.INVALID_CONTENT_TYPE;
    ExceptionLogger.warn(request, errorCode, exception);
    return ResponseEntity.status(errorCode.getStatus())
        .body(ErrorResponse.from(errorCode));
  }

  @ExceptionHandler
  public ResponseEntity<ErrorResponse> handleMethodArgumentNotValid(
      HttpServletRequest request,
      MethodArgumentNotValidException exception
  ) {
    ErrorCode errorCode = ErrorCode.CLIENT_BAD_REQUEST;
    String fieldErrorMessage = exception.getBindingResult()
        .getFieldErrors()
        .getFirst()
        .getDefaultMessage();
    String message = Objects.requireNonNullElse(fieldErrorMessage, errorCode.getMessage());
    ExceptionLogger.warn(request, errorCode, message, exception);
    return ResponseEntity.status(errorCode.getStatus())
        .body(ErrorResponse.of(errorCode, message));
  }

  @ExceptionHandler(produces = MediaType.APPLICATION_JSON_VALUE)
  public ResponseEntity<ErrorResponse> handleAuthentication(
      HttpServletRequest request,
      AuthenticationException exception
  ) {
    ErrorCode errorCode = exception.getErrorCode();
    ExceptionLogger.info(request, errorCode, exception);
    return ResponseEntity.status(errorCode.getStatus())
        .body(ErrorResponse.from(errorCode));
  }

  @ExceptionHandler(produces = MediaType.TEXT_EVENT_STREAM_VALUE)
  public ResponseEntity<String> handleAuthenticationForSse(
      HttpServletRequest request,
      AuthenticationException exception
  ) {
    ErrorCode errorCode = exception.getErrorCode();
    ExceptionLogger.info(request, errorCode, exception);
    return ResponseEntity.status(errorCode.getStatus())
        .contentType(MediaType.TEXT_EVENT_STREAM)
        .body("data: " + errorCode.getMessage() + "\n\n");
  }

  @ExceptionHandler(produces = MediaType.APPLICATION_JSON_VALUE)
  public ResponseEntity<ErrorResponse> handleAccessDenied(
      HttpServletRequest request,
      AccessDeniedException exception
  ) {
    ErrorCode errorCode = exception.getErrorCode();
    ExceptionLogger.info(request, errorCode, exception);
    return ResponseEntity.status(errorCode.getStatus())
        .body(ErrorResponse.from(errorCode));
  }

  @ExceptionHandler(produces = MediaType.TEXT_EVENT_STREAM_VALUE)
  public ResponseEntity<String> handleAccessDeniedForSse(
      HttpServletRequest request,
      AccessDeniedException exception
  ) {
    ErrorCode errorCode = exception.getErrorCode();
    ExceptionLogger.info(request, errorCode, exception);
    return ResponseEntity.status(errorCode.getStatus())
        .contentType(MediaType.TEXT_EVENT_STREAM)
        .body("data: " + errorCode.getMessage() + "\n\n");
  }

  @ExceptionHandler
  public ResponseEntity<ErrorResponse> handleBusiness(
      HttpServletRequest request,
      BusinessException exception
  ) {
    ErrorCode errorCode = exception.getErrorCode();
    ExceptionLogger.warn(request, errorCode, exception);
    return ResponseEntity.status(errorCode.getStatus())
        .body(ErrorResponse.from(errorCode));
  }

  @ExceptionHandler
  public ResponseEntity<ErrorResponse> handleFeign(
      HttpServletRequest request,
      FeignException exception
  ) {
    ErrorCode errorCode = ErrorCode.FAILED_EXTERNAL_SERVER_COMMUNICATION;
    String message = exception.contentUTF8();
    ExceptionLogger.warn(request, errorCode, message, exception);
    return ResponseEntity.status(errorCode.getStatus())
        .body(ErrorResponse.from(errorCode));
  }

  @ExceptionHandler
  public ResponseEntity<ErrorResponse> handleResourceNotFound(
      HttpServletRequest request,
      NoResourceFoundException exception
  ) {
    ErrorCode errorCode = ErrorCode.RESOURCE_NOT_FOUND;
    ExceptionLogger.warn(request, errorCode, exception);
    return ResponseEntity.status(errorCode.getStatus())
        .body(ErrorResponse.from(errorCode));
  }

  @ExceptionHandler
  public ResponseEntity<ErrorResponse> handleMethodNotAllowed(
      HttpServletRequest request,
      HttpRequestMethodNotSupportedException exception
  ) {
    ErrorCode errorCode = ErrorCode.METHOD_NOT_ALLOWED;
    String message = errorCode.getMessage()
        .formatted(request.getRequestURI(), exception.getMethod());
    ExceptionLogger.warn(request, errorCode, message, exception);
    return ResponseEntity.status(errorCode.getStatus())
        .body(ErrorResponse.of(errorCode, message));
  }

  @ExceptionHandler(AsyncRequestTimeoutException.class)
  public ResponseEntity<String> handleAsyncRequestTimeout() {
    return ResponseEntity
        .status(HttpStatus.REQUEST_TIMEOUT)
        .contentType(MediaType.TEXT_EVENT_STREAM)
        .body("data: CLOSED\n\n");
  }

  @ExceptionHandler(produces = MediaType.APPLICATION_JSON_VALUE)
  public ResponseEntity<ErrorResponse> handleServer(
      HttpServletRequest request,
      Exception exception
  ) {
    ErrorCode errorCode = ErrorCode.INTERNAL_SERVER_ERROR;
    String message = Objects.requireNonNullElse(exception.getMessage(), errorCode.getMessage());
    if (!message.equalsIgnoreCase("Broken pipe")) {
      ExceptionLogger.error(request, errorCode, exception);
    }
    return ResponseEntity.internalServerError()
        .body(ErrorResponse.from(errorCode));
  }

  @ExceptionHandler(produces = MediaType.TEXT_EVENT_STREAM_VALUE)
  public ResponseEntity<String> handleServerForSse(
      HttpServletRequest request,
      Exception exception
  ) {
    ErrorCode errorCode = ErrorCode.INTERNAL_SERVER_ERROR;
    String message = Objects.requireNonNullElse(exception.getMessage(), errorCode.getMessage());
    if (!message.equalsIgnoreCase("Broken pipe")) {
      ExceptionLogger.error(request, errorCode, exception);
    }
    return ResponseEntity.internalServerError()
        .contentType(MediaType.TEXT_EVENT_STREAM)
        .body("data: " + message + "\n\n");
  }

}

File: src/main/kotlin/com/everyonewaiter/global/exception/GlobalExceptionHandler.kt
package com.everyonewaiter.global.exception

import io.github.oshai.kotlinlogging.KotlinLogging
import jakarta.servlet.http.HttpServletRequest
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.web.HttpMediaTypeNotSupportedException
import org.springframework.web.HttpRequestMethodNotSupportedException
import org.springframework.web.bind.MissingServletRequestParameterException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException
import org.springframework.web.servlet.resource.NoResourceFoundException

private val logger = KotlinLogging.logger {}

@RestControllerAdvice
internal class GlobalExceptionHandler {
    @ExceptionHandler(value = [IllegalArgumentException::class, IllegalStateException::class])
    fun handleIllegalException(
        request: HttpServletRequest,
        exception: Exception,
    ): ResponseEntity<ErrorResponse> {
        val errorCode = ErrorCode.BAD_REQUEST
        val message = exception.message ?: errorCode.message
        logWarn(request, errorCode, message, exception)
        return ResponseEntity.status(errorCode.status).body(ErrorResponse(errorCode, message))
    }

    @ExceptionHandler
    fun handleMethodArgsTypeMismatchException(
        request: HttpServletRequest,
        exception: MethodArgumentTypeMismatchException,
    ): ResponseEntity<ErrorResponse> {
        val errorCode = ErrorCode.INVALID_PARAMETER
        logWarn(request, errorCode, errorCode.message, exception)
        return ResponseEntity.status(errorCode.status).body(ErrorResponse(errorCode, errorCode.message))
    }

    @ExceptionHandler
    fun handleMissingServletRequestParameterException(
        request: HttpServletRequest,
        exception: MissingServletRequestParameterException,
    ): ResponseEntity<ErrorResponse> {
        val errorCode = ErrorCode.MISSING_PARAMETERS
        logWarn(request, errorCode, errorCode.message, exception)
        return ResponseEntity.status(errorCode.status).body(ErrorResponse(errorCode, errorCode.message))
    }

    @ExceptionHandler
    fun handleHttpMessageNotReadableException(
        request: HttpServletRequest,
        exception: HttpMessageNotReadableException,
    ): ResponseEntity<ErrorResponse> {
        val errorCode = ErrorCode.INVALID_REQUEST_BODY
        logWarn(request, errorCode, errorCode.message, exception)
        return ResponseEntity.status(errorCode.status).body(ErrorResponse(errorCode, errorCode.message))
    }

    @ExceptionHandler
    fun handleHttpMediaTypeNotSupportedException(
        request: HttpServletRequest,
        exception: HttpMediaTypeNotSupportedException,
    ): ResponseEntity<ErrorResponse> {
        val errorCode = ErrorCode.INVALID_CONTENT_TYPE
        logWarn(request, errorCode, errorCode.message, exception)
        return ResponseEntity.status(errorCode.status).body(ErrorResponse(errorCode, errorCode.message))
    }

    @ExceptionHandler
    fun handleBusinessException(
        request: HttpServletRequest,
        exception: BusinessException,
    ): ResponseEntity<ErrorResponse> {
        val errorCode = exception.errorCode
        logWarn(request, errorCode, errorCode.message, exception)
        return ResponseEntity.status(errorCode.status).body(ErrorResponse(errorCode, errorCode.message))
    }

    @ExceptionHandler
    fun handleAuthenticationException(
        request: HttpServletRequest,
        exception: AuthenticationException,
    ): ResponseEntity<ErrorResponse> {
        val errorCode = exception.errorCode
        if (errorCode == ErrorCode.UNAUTHORIZED) {
            logInfo(request, errorCode, exception.message, exception)
        } else {
            logWarn(request, errorCode, exception.message, exception)
        }
        return ResponseEntity.status(errorCode.status).body(ErrorResponse(errorCode, exception.message))
    }

    @ExceptionHandler
    fun handleAccessDeniedException(
        request: HttpServletRequest,
        exception: AccessDeniedException,
    ): ResponseEntity<ErrorResponse> {
        val errorCode = exception.errorCode
        if (errorCode == ErrorCode.FORBIDDEN) {
            logInfo(request, errorCode, exception.message, exception)
        } else {
            logWarn(request, errorCode, exception.message, exception)
        }
        return ResponseEntity.status(errorCode.status).body(ErrorResponse(errorCode, exception.message))
    }

    @ExceptionHandler
    fun handleResourceNotFoundException(
        request: HttpServletRequest,
        exception: NoResourceFoundException,
    ): ResponseEntity<ErrorResponse> {
        val errorCode = ErrorCode.RESOURCE_NOT_FOUND
        logWarn(request, errorCode, errorCode.message, exception)
        return ResponseEntity.status(errorCode.status).body(ErrorResponse(errorCode, errorCode.message))
    }

    @ExceptionHandler
    fun handleMethodNotAllowedException(
        request: HttpServletRequest,
        exception: HttpRequestMethodNotSupportedException,
    ): ResponseEntity<ErrorResponse> {
        val errorCode = ErrorCode.METHOD_NOT_ALLOWED
        val message = errorCode.message.format(request.requestURI, exception.method)
        logWarn(request, errorCode, message, exception)
        return ResponseEntity.status(errorCode.status).body(ErrorResponse(errorCode, message))
    }

    @ExceptionHandler
    fun handleInternalServerException(
        request: HttpServletRequest,
        exception: Exception,
    ): ResponseEntity<ErrorResponse> {
        val errorCode = ErrorCode.INTERNAL_SERVER_ERROR
        if (exception.message != "Broken pipe") {
            logError(request, exception)
        }
        return ResponseEntity.internalServerError().body(ErrorResponse(code = errorCode, message = errorCode.message))
    }

    private fun logInfo(
        request: HttpServletRequest,
        errorCode: ErrorCode,
        message: String,
        exception: Exception,
    ) {
        val method = request.method
        val uri = request.requestURI
        val exceptionName = exception.javaClass.simpleName
        logger.info { "[${errorCode.name}] [$method $uri] [$exceptionName ${errorCode.status}]: $message" }
    }

    private fun logWarn(
        request: HttpServletRequest,
        errorCode: ErrorCode,
        message: String,
        exception: Exception,
    ) {
        val method = request.method
        val uri = request.requestURI
        val exceptionName = exception.javaClass.simpleName
        logger.warn { "[${errorCode.name}] [$method $uri] [$exceptionName ${errorCode.status}]: $message" }
    }

    private fun logError(
        request: HttpServletRequest,
        exception: Exception,
    ) {
        val method = request.method
        val uri = request.requestURI
        val status = HttpStatus.INTERNAL_SERVER_ERROR
        val exceptionName = exception.javaClass.simpleName
        logger.error(exception) { "[$method $uri] [$exceptionName $status]: ${exception.message}" }
    }
}
